---
title: "The HCAExplorer Package"
author: "Daniel Van Twisk"
date: "`r format(Sys.Date(), '%A, %B %d, %Y')`"
always_allow_html: yes
output:
  BiocStyle::html_document:
    df_print: paged
    toc_float: true
abstract: >
  The [Human Cell Atlas] (HCA) (https://www.humancellatlas.org/) was created in
  order to create comprehensive reference maps of all human cells as a basis for
  both understanding human health and diagnosing, monitoring, and treating
  disease. The *HCAExplorer* Biocondctor pacakge provides infrastructure for
  searching for, queerying, and accessing data help on the Human Cell Atlas's
  Data Coordination Platform (https://dss.data.humancellatlas.org/). Further
  changes to the package are planned to incorperate higer level functionality to
  upload user generated data to the the Human Cell Atlas platform.
vignette: >
  %\VignetteIndexEntry{Introduction to Accessing the HCAExplorer using the
        HCAExplorer package}
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{HCAExplorer}
  %\VignetteEncoding{UTF-8}
---

```{r init, results='hide', echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
opts_chunk$set(warning=FALSE, message=FALSE)
BiocStyle::markdown()
```


# What is the Human Cell Atlas?

From the [Human Cell Atlas (HCA) website](https://www.humancellatlas.org/):

> The cell is the core unit of the human body—the key to understanding the
biology of health and the ways in which molecular dysfunction leads to disease.
Yet our characterization of the hundreds of types and subtypes of cells in the
human body is limited, based partly on techniques that have limited resolution
and classifications that do not always map neatly to each other. Genomics has
offered a systematic approach, but it has largely been applied in bulk to many
cell types at once—masking critical differences between cells—and in isolation
from other valuable sources of data.

> Recent advances in single-cell genomic analysis of cells and tissues have put
systematic, high-resolution and comprehensive reference maps of all human cells
within reach. In other words, we can now realistically envision a human cell
atlas to serve as a basis for both understanding human health and diagnosing,
monitoring, and treating disease.

> At its core, a cell atlas would be a collection of cellular reference maps,
characterizing each of the thousands of cell types in the human body and where
they are found. It would be an extremely valuable resource to empower the global
research community to systematically study the biological changes associated
with different diseases, understand where genes associated with disease are
active in our bodies, analyze the molecular  mechanisms that govern the
production and activity of different cell types, and sort out how different cell
types combine and work together to form tissues.

The Human Cell Atlas facilitates queries on it's [data coordination platform with
a RESTFUL API] (https://dss.data.humancellatlas.org/).

## Installation

To install this package, use Bioconductor's `BiocManager` package.

```{r install_bioc, eval=FALSE}
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install('HCAExplorer')
```

```{r libraries, message=FALSE}
library(HCAExplorer)
```

## Connecting to the Human Cell Atlas

The `r Biocpkg("HCAExplorer")` package relies on having network
connectivety. Also, the a link to a viable digest of the Human Cell Atlas must
also be operational. The backend that we are using will be using is refered to
as the "azul backend". This package is meant to mirror the functionality of the
[HCA Data Explorer](https://data.humancellatlas.org/explore/projects).

The `HCAExplorer` object serves as the representation of the Human Cell
Atlas. Upon creation, it will automatically perform a cursorary query and
display a small table showing the first few project of the entire HCA. This
intial table contains some columns that we have determined are most useful
to users. The output also displays the url of the instance of the HCA digest
being used, the current query, relevant information about the quantity of data
being displayed, and finally the table of projects.

By default, 15 bundles per page will be displayed in the result and the
default url to the HCA DCP will be used. These two values can be changed in
the constructor or later on using methods.

If the HCA cannot be reached, an error will be thrown displaying the status of
the request.

```{r createHCA}
hca <- HCAExplorer(url = 'https://service.explore.data.humancellatlas.org', per_page = 15)
hca
```

Upon displaying the object, multiple fields can be seen:
- The class: `HCAExplorer`
- The azul-backend address that is currently being used.
- The current query.
- The projects being shown and whether a `link` to more results is availiable.
- The number of projects being shown per_page.
- The results `tibble` of the query. (This table is abbreviated to show only
  columns that we determined are most useful to the user.)

The results `tibble` can be obtained using the `results()` method.

```{r results}
results(hca)
```

To toggle whether prjects, samples, or file are being displayed in the `tibble`,
the `activate()` method can be used to choose which to display.

```{r activate}
## Bundles are diaplyed be default
nrow(results(hca))

## The HCAExplorer object is activated here by 'files'
hca <- hca %>% activate('samples')
hca
nrow(results(hca))

## Revert back to showing bundles with 'bundles'
hca <- hca %>% activate('projects')
```

Since there are more projects in the HCA than can be shown, if `link` is
`True`, the next set of bundles can be obtained using the `nextResults` method.

```{r nextResults}
hca <- nextResults(hca)
hca
```

# A working example: Obtaining Metadata files from the HCAExplorer

One of the primary tasks of the HCAExplorer is to obtain metadata files of
projects and then pass them down to other pipelines to download useful
information like expression matrices.

## Querying the HCAExplorer

To show which fields are available to query, use the `supportedFilters()` method.

```{r fields}
hca <- HCAExplorer()
hca %>% fields
```

The `abbreviated_fields` column indicates the shortest possible name that can be
used to reference the field. The `field_names` columns shows the whole schema
name on the json schema.

Availiable values to these fields can be found using the `values()` method.
If the parameter `fields` is specified, it will display values related to those
field. If not specified, the values for all fields will be shown.

```{r valuess}
hca %>% values(c('organ'))
```

HCAExplorer extends the functionality of the `r CRANpkg("dplyr")` package's `filter()`
and `select()` methods.

The `filter()` method allows the user to query the HCA by relating fields to
certain values. Character fields can be queried using the operators:
- `==`
- `%in%`

Combination operators can be used to combine queries
- `&`

As an example, in order to find HCA resources associated with the brain. It can
the be seen by looking at the result of the `fields` method, that
`organ.text` can be used to reference `files.specimen_from_organism.organ.text`.
Running `organ.text` through `values`, we see the following:

```{r availabeOrgan}
hca %>% values('organ')
```

Now we see that "brain" and "Brain" are available values. Since these values are
the result of input by other users, there may be errors or inconsistencies. To
be safe, both fields can be queried with the following query:

```{r firstFilter}
hca2 <- hca %>% filter(organ == c('blood', 'brain'))
hca2 <- hca %>% filter(organ %in% c('blood', 'brain'))
hca2
```

If we also wish to search for results based on the NCBI Taxon ID for mouse,
10090, as well as brain, we can perform this query in a variety of ways.
```{r multiFilter}
hca2 <- hca %>% filter(organ %in% c('Brain', 'brain')) %>% filter(disease == 'normal')
hca2 <- hca %>% filter(organ %in% c('Brain', 'brain'), disease == 'normal')
hca <- hca %>% filter(organ %in% c('Brain', 'brain') & disease == 'normal')
hca
```

The `HCAExplorer` object can undo the most recent queries run on it.

```{r undoQuery}
hca <- HCAExplorer()
hca <- hca %>% filter(organ == blood)
hca <- hca %>% filter(disease == normal)
hca <- hca %>% undoQuery(n = 1L)
hca
```

If one would want to start from a fresh query but retain the modifications made
to the `HCAExplorer` object, the `resetEsQuery()` method can be used.

```
hca <- hca %>% resetQuery
hca
```

```{r sessionInfo}
sessionInfo()
```

# Developer notes

- The `S3` object-oriented programming paradigm is used.
- Methods from the `dplyr` package can be used to manipulate objects in the
`HCAExplorer` package.
- In the future, we wish to expand the functionalit of this packages to cover
the remaining functionality of the hca dcp api.

